pkg/api/pod/warnings.go

```diff
@@ -0,0 +1,240 @@
+			}
+		}
+	}
+	// imagePullSecrets with empty name (#99454#issuecomment-787838112)
+	for i, item := range podSpec.ImagePullSecrets {
+		if len(item.Name) == 0 {
+			warnings = append(warnings, fmt.Sprintf("%s: invalid empty name %q", fieldPath.Child("spec", "imagePullSecrets").Index(i).Child("name"), item.Name))
+		}
+	}
+
+	// duplicate volume names (#78266, #58477)
+	if len(podSpec.Volumes) > 1 {
+		items := sets.NewString()
+		for i, item := range podSpec.Volumes {
+			if items.Has(item.Name) {
+				warnings = append(warnings, fmt.Sprintf("%s: duplicate name %q", fieldPath.Child("spec", "volumes").Index(i).Child("name"), item.Name))
+			} else {
+				items.Insert(item.Name)
+			}
+		}
+	}
+
+	// fractional memory/ephemeral-storage requests/limits (#79950, #49442, #18538)
+	if value, ok := podSpec.Overhead[api.ResourceMemory]; ok && value.MilliValue()%int64(1000) != int64(0) {
+		warnings = append(warnings, fmt.Sprintf("%s: fractional byte value %q is invalid, must be an integer", fieldPath.Child("spec", "overhead").Key(string(api.ResourceMemory)), value.String()))
+	}
+	if value, ok := podSpec.Overhead[api.ResourceEphemeralStorage]; ok && value.MilliValue()%int64(1000) != int64(0) {
+		warnings = append(warnings, fmt.Sprintf("%s: fractional byte value %q is invalid, must be an integer", fieldPath.Child("spec", "overhead").Key(string(api.ResourceEphemeralStorage)), value.String()))
+	}
+
+	pods.VisitContainersWithPath(podSpec, fieldPath.Child("spec"), func(c *api.Container, p *field.Path) bool {
+		// fractional memory/ephemeral-storage requests/limits (#79950, #49442, #18538)
+		if value, ok := c.Resources.Limits[api.ResourceMemory]; ok && value.MilliValue()%int64(1000) != int64(0) {
+			warnings = append(warnings, fmt.Sprintf("%s: fractional byte value %q is invalid, must be an integer", p.Child("resources", "limits").Key(string(api.ResourceMemory)), value.String()))
+		}
+		if value, ok := c.Resources.Requests[api.ResourceMemory]; ok && value.MilliValue()%int64(1000) != int64(0) {
+			warnings = append(warnings, fmt.Sprintf("%s: fractional byte value %q is invalid, must be an integer", p.Child("resources", "requests").Key(string(api.ResourceMemory)), value.String()))
+		}
+		if value, ok := c.Resources.Limits[api.ResourceEphemeralStorage]; ok && value.MilliValue()%int64(1000) != int64(0) {
+			warnings = append(warnings, fmt.Sprintf("%s: fractional byte value %q is invalid, must be an integer", p.Child("resources", "limits").Key(string(api.ResourceEphemeralStorage)), value.String()))
+		}
+		if value, ok := c.Resources.Requests[api.ResourceEphemeralStorage]; ok && value.MilliValue()%int64(1000) != int64(0) {
+			warnings = append(warnings, fmt.Sprintf("%s: fractional byte value %q is invalid, must be an integer", p.Child("resources", "requests").Key(string(api.ResourceEphemeralStorage)), value.String()))
+		}
+
+		// duplicate containers[*].env (#86163, #93266, #58477)
+		if len(c.Env) > 1 {
+			items := sets.NewString()
```

Comment:
While it would be uglier, it would be better to reuse one string set on this entire function and reset before use.

Reply:
how much better? benchmarks didn't look that concerning to me

```
BenchmarkNoWarnings-12    	  491527	      2534 ns/op	     336 B/op	       7 allocs/op
```

clearing and reusing that set between blocks actually made the benchmark worse:

```
BenchmarkNoWarnings-12    	  521125	      2552 ns/op	     528 B/op	       9 allocs/op
```

Reply:
I don't know how clearing and reusing could make it worse - how did you clear?

Reply:
This is a bug in the set - it should have a Clear() or Reset() method that does `for k := range m { delete(m, k) }` which the go compiler optimizes to "clear but reuse buckets", which is roughly free

Reply:
hmm... reworked my clearing, dropped it back down to the original, but no cheaper than a new set per block:

```diff
+func clear(items sets.String) {
+       for k := range items {
+               delete(items, k)
+       }
+}
+
 func warningsForPodSpecAndMeta(fieldPath *field.Path, podSpec *api.PodSpec, meta *metav1.ObjectMeta, oldPodSpec *api.PodSpec, oldMeta
        var warnings []string
+       var items sets.String

        // use of deprecated node labels in selectors/affinity/topology
        for k := range podSpec.NodeSelector {
@@ -159,7 +166,11 @@ func warningsForPodSpecAndMeta(fieldPath *field.Path, podSpec *api.PodSpec, meta

        // duplicate hostAliases (#91670, #58477)
        if len(podSpec.HostAliases) > 1 {
-               items := sets.NewString()
+               if items == nil {
+                       items = sets.NewString()
+               } else {
+                       clear(items)
+               }
```

```
BenchmarkNoWarnings-12    	  702271	      1959 ns/op	     336 B/op	       7 allocs/op
```

Reply:
is there another way I should be benchmarking to demonstrate the value of the increased ugliness?

Reply:
Are you comfortable with me getting a rain check to come back later and frown at the impact it has and volunteering to review any slight ugliness I add later?  I'm comfortable deferring since micro didn't improve.